lib/misc.js

```diff
@@ -63,5 +65,91 @@ exports.generateId = function(fields) {
   hasher.update('_');
 
   return base32hex.encodeBuffer(hasher.digest(), { paddingChar: '-' });
-};
+}
+
+function generateInternalId(externalId, groupId) {
+  return generateId([externalId, groupId]);
+}
+
+function ensureInternalId(datum) {
+  if (!datum.id) {
+    datum.id = generateExternalId(datum);
+  }
+  if (!datum._id) {
+    datum._id = generateInternalId(datum.id, datum._groupId);
+  }
+  return datum;
+}
+
+function generateExternalId(datum) {
+  return schema.generateId(datum, schema.idFields(datum.type));
+}
+
+function sortTimeAsc(a, b) {
+  const timeA = new Date(a.time);
+  const timeB = new Date(b.time);
+  if (timeA < timeB)
+    return -1;
+  if (timeA > timeB)
+    return 1;
+  return 0;
+}
+
+/**
+ * @typedef {Object} RemoveDuplicatesResult
+ * @property {Object[]} array - the non duplicates that should be persisted to the database.
+ * @property {Object[]} duplicates - duplicate array entries that should not be persisted (or would return an error if they would).
+ * @property {Object} latestData - a convenience mapping of [_id] => datum
+ */
+
+/**
+ * Returns data that should be inserted into the database taking into account existing data.
+ *
+ * @param {Object[]} existingDBData - Array of data already existing in the database.
+ * @param {Object[]} inputData - Array of data that user intends to save to the database.
+ * @return {RemoveDuplicatesResult}
+ */
+function removeDuplicates(existingDBData, inputData) {
+  const latestData = {};
+  for (const datum of existingDBData) {
+    const externalId = generateExternalId(datum);
+    const internalId = generateInternalId(externalId, datum._groupId);
+    latestData[datum._id] = datum;
+  }
+
+  // Duplicate inserted data may exist in the inputData array. This occurs
+  // when there is multiple data with the same id / _id. This is not allowed
+  // and in the serial code would return a { statusCode: 400,
+  // errorCode: 'duplicate', message: 'received a duplicate event'} error to
+  // the callback. Since there may be multiple duplicates we instead include
+  // an array of duplicates in the returned object.
+  const finalArray = [];
+  const duplicates = [];
+  inputData = inputData.slice(0);
+  inputData.sort(sortTimeAsc);
+  for (const datum of inputData) {
+    const externalId = generateExternalId(datum);
+    const internalId = generateInternalId(externalId, datum._groupId);
+
+    // The absence of datum.createdTime indicates an insert whereas the
+    // presence of it indicates an update. Disallow multiple inserts.
```

Comment:
Is this true? That is, that the existence of `datum.createdtime` indicates that the incoming datum is intended to update existing datum? My memory is fuzzy in this area. I suppose Gerrit would know.

Reply:
That is my understanding of how it works. @gniezen can you confirm?

Reply:
I'm not aware of any updates that we do for Jellyfish records - my understanding has always been that Jellyfish will drop the record on the floor if it already exists. Never knew about this "workaround"  to update records! :exploding_head: